charmcraft/parts/plugins/_poetry.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2024 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+"""Charmcraft-specific poetry plugin."""
+
+import pathlib
+from pathlib import Path
+
+from craft_parts.plugins import poetry_plugin
+from overrides import override
+
+from charmcraft import utils
+
+
+class PoetryPluginProperties(poetry_plugin.PoetryPluginProperties, frozen=True):
+
+    poetry_keep_bins: bool = False
+    """Keep the virtual environment's 'bin' directory."""
+
+
+class PoetryPlugin(poetry_plugin.PoetryPlugin):
+    """Charmcraft-specific version of the poetry plugin."""
+
+    properties_class = PoetryPluginProperties
+    _options: PoetryPluginProperties  # type: ignore[reportIncompatibleVariableOverride]
+
+    def get_build_environment(self) -> dict[str, str]:
+        return utils.extend_python_build_environment(super().get_build_environment())
+
+    def _get_venv_directory(self) -> Path:
+        return self._part_info.part_install_dir / "venv"
+
+    def _get_pip(self) -> str:
+        """Get the pip command to use."""
+        return f"{self._get_system_python_interpreter()} -m pip --python=${{PARTS_PYTHON_VENV_INTERP_PATH}}"
+
+    def _get_pip_install_commands(self, requirements_path: pathlib.Path) -> list[str]:
+        """Get the commands for installing with pip.
+
+        This only installs the dependencies from requirements, unlike the upstream version.
```

Comment:
how come?

Reply:
Because charms are not Python packages, so `pip install .` won't work.

Reply:
I've made the relevant docstrings both here and in the python module more explicit to explain this.